### ISSUE TITLE
Problem in making prediction

### ISSUE THREAD

**Issue by xuehy** [pluskid/Mocha.jl#132]
Hi! I have trained a network with data layer of type HDF5DataLayer. Now I want to make predictions for images and I build a MemoryDataLayer and replace the network's HDF5DataLayer with the new layer as well as the loss layer with SoftmaxLayer.

However, when I tried to load snapshot of the existing network into the new net,
```
load_snapshot(net,"snapshot")
```
 there occurred an error

```
DEBUG:root:Loading parameters for layer conv1
ERROR: Not implemented (should copy content of src to dst)
```
This is the network for training.

```
data_layer  = AsyncHDF5DataLayer(name="train-data", source="data/train.txt", batch_size=64, shuffle=true)
conv_layer  = ConvolutionLayer(name="conv1", n_filter=20, kernel=(5,5), bottoms=[:data], tops=[:conv])
pool_layer  = PoolingLayer(name="pool1", kernel=(2,2), stride=(2,2), bottoms=[:conv], tops=[:pool])
conv2_layer = ConvolutionLayer(name="conv2", n_filter=50, kernel=(5,5), bottoms=[:pool], tops=[:conv2])
pool2_layer = PoolingLayer(name="pool2", kernel=(2,2), stride=(2,2), bottoms=[:conv2], tops=[:pool2])
fc1_layer   = InnerProductLayer(name="ip1", output_dim=500, neuron=Neurons.ReLU(), bottoms=[:pool2], tops=[:ip1])
fc2_layer   = InnerProductLayer(name="ip2", output_dim=10, bottoms=[:ip1], tops=[:ip2])
loss_layer  = SoftmaxLossLayer(name="loss", bottoms=[:ip2,:label])

```

This is the network I build for forwarding prediction.
```
data_layer  = MemoryDataLayer(name="data",tops=[:data],batch_size=1,data = Array[zeros(28,28,1,1)])
conv_layer  = ConvolutionLayer(name="conv1", n_filter=20, kernel=(5,5), bottoms=[:data], tops=[:conv])
pool_layer  = PoolingLayer(name="pool1", kernel=(2,2), stride=(2,2), bottoms=[:conv], tops=[:pool])
conv2_layer = ConvolutionLayer(name="conv2", n_filter=50, kernel=(5,5), bottoms=[:pool], tops=[:conv2])
pool2_layer = PoolingLayer(name="pool2", kernel=(2,2), stride=(2,2), bottoms=[:conv2], tops=[:pool2])
fc1_layer   = InnerProductLayer(name="ip1", output_dim=500, neuron=Neurons.ReLU(), bottoms=[:pool2], tops=[:ip1])
fc2_layer   = InnerProductLayer(name="ip2", output_dim=10, bottoms=[:ip1], tops=[:ip2])
softmax_layer = SoftmaxLayer(name="prob", bottoms=[:ip2], tops=[:prob])
```
How can I solve this problem?

Thanks!

**Comment by pluskid** [comment]
Is the images you used to train and test the network of the same shape? Basically, the input shape will define the shape of the following layers in the network. Can you `print` the net constructed for both testing and training to see if they have the same shape of input-output blobs in each layer?

**Comment by xuehy** [comment]
training network:

```

************************************************************
          NAME: MNIST-train
       BACKEND: CPUBackend
  ARCHITECTURE: 8 layers
............................................................
 *** AsyncHDF5DataLayer(train-data)
    Outputs ---------------------------
          data: Blob(28 x 28 x 1 x 64)
         label: Blob(1 x 64)
............................................................
 *** ConvolutionLayer(conv1)
    Inputs ----------------------------
          data: Blob(28 x 28 x 1 x 64)
    Outputs ---------------------------
          conv: Blob(24 x 24 x 20 x 64)
............................................................
 *** PoolingLayer(pool1)
    Inputs ----------------------------
          conv: Blob(24 x 24 x 20 x 64)
    Outputs ---------------------------
          pool: Blob(12 x 12 x 20 x 64)
............................................................
 *** ConvolutionLayer(conv2)
    Inputs ----------------------------
          pool: Blob(12 x 12 x 20 x 64)
    Outputs ---------------------------
         conv2: Blob(8 x 8 x 50 x 64)
............................................................
 *** PoolingLayer(pool2)
    Inputs ----------------------------
         conv2: Blob(8 x 8 x 50 x 64)
    Outputs ---------------------------
         pool2: Blob(4 x 4 x 50 x 64)
............................................................
 *** InnerProductLayer(ip1)
    Inputs ----------------------------
         pool2: Blob(4 x 4 x 50 x 64)
    Outputs ---------------------------
           ip1: Blob(500 x 64)
............................................................
 *** InnerProductLayer(ip2)
    Inputs ----------------------------
           ip1: Blob(500 x 64)
    Outputs ---------------------------
           ip2: Blob(10 x 64)
............................................................
 *** SoftmaxLossLayer(loss)
    Inputs ----------------------------
           ip2: Blob(10 x 64)
         label: Blob(1 x 64)
************************************************************
```

testing:

```
************************************************************
          NAME: MNIST-train
       BACKEND: CPUBackend
  ARCHITECTURE: 8 layers
............................................................
 *** MemoryDataLayer(data)
    Outputs ---------------------------
          data: Blob(28 x 28 x 1 x 1)
............................................................
 *** ConvolutionLayer(conv1)
    Inputs ----------------------------
          data: Blob(28 x 28 x 1 x 1)
    Outputs ---------------------------
          conv: Blob(24 x 24 x 20 x 1)
............................................................
 *** PoolingLayer(pool1)
    Inputs ----------------------------
          conv: Blob(24 x 24 x 20 x 1)
    Outputs ---------------------------
          pool: Blob(12 x 12 x 20 x 1)
............................................................
 *** ConvolutionLayer(conv2)
    Inputs ----------------------------
          pool: Blob(12 x 12 x 20 x 1)
    Outputs ---------------------------
         conv2: Blob(8 x 8 x 50 x 1)
............................................................
 *** PoolingLayer(pool2)
    Inputs ----------------------------
         conv2: Blob(8 x 8 x 50 x 1)
    Outputs ---------------------------
         pool2: Blob(4 x 4 x 50 x 1)
............................................................
 *** InnerProductLayer(ip1)
    Inputs ----------------------------
         pool2: Blob(4 x 4 x 50 x 1)
    Outputs ---------------------------
           ip1: Blob(500 x 1)
............................................................
 *** InnerProductLayer(ip2)
    Inputs ----------------------------
           ip1: Blob(500 x 1)
    Outputs ---------------------------
           ip2: Blob(10 x 1)
............................................................
 *** SoftmaxLayer(prob)
    Inputs ----------------------------
           ip2: Blob(10 x 1)
    Outputs ---------------------------
          prob: Blob(10 x 1)
************************************************************

```

I notice that the ```batch_size``` are not the same. Is this the problem? How can I build a testing network that takes only one image at a time?

**Comment by xuehy** [comment]
Problem solved!

``` load_network``` works differently as ``` load_snapshot```. ```load_network``` needs the network to be completely the same as the one saved in files while the latter allows different data layer.

Sorry for reporting a false bug.

**Comment by slundberg** [comment]
Can I ask how you solved this? It seems natural to use `load_snapshot` (which just calls `load_network` inside).

I get the same error you describe above and it comes from the function called in [src/utils/io.jl#L103](https://github.com/pluskid/Mocha.jl/blob/5451ed19d526d1b20b644653800df394c6d9d555/src/utils/io.jl#L103) not being implemented.

**Comment by slundberg** [comment]
In case anyone else has the same issue, the problem is that the HDF5 dataset is `Float32` while the MemoryDataLayer created above is `Float64` (since that is the default type of `zeros`).
